src/Tools/dotnet-monitor/EndpointInfo/ServerEndpointInfoSource.cs

```diff
@@ -248,33 +138,130 @@ private async Task ResumeAndQueueEndpointInfo(IpcEndpointInfo info, Cancellation
+                await _pendingRemovalEndpointsSemaphore.WaitAsync(token).ConfigureAwait(false);
+
+                IEndpointInfo endpoint;
+                lock (_pendingRemovalEndpoints)
+                {
+                   endpoint = _pendingRemovalEndpoints.Dequeue();
+                }
+
+                foreach (IEndpointInfoSourceCallbacks callback in _callbacks)
+                {
+                    await callback.OnRemovedEndpointInfoAsync(endpoint, token).ConfigureAwait(false);
+                }
+
+                server.RemoveConnection(endpoint.RuntimeInstanceCookie);
             }
         }
 
-        private void VerifyIsListening()
+        private async Task PruneEndpointsAsync(List<IEndpointInfo> validEndpoints, CancellationToken token)
         {
-            if (!IsListening)
+            List<IEndpointInfo> endpointsToRemove = new();
+
+            // Prune connections that no longer have an active runtime instance before
+            // returning the list of connections.
+            await _activeEndpointsSemaphore.WaitAsync(token).ConfigureAwait(false);
+
+            try
+            {
+                // Check the transport for each endpoint info and remove it if the check fails.
+                List<Task<bool>> checkTasks = new();
+                foreach (EndpointInfo info in _activeEndpoints)
+                {
+                    checkTasks.Add(Task.Run(() => CheckEndpointAsync(info, token), token));
+                }
+
+                // Wait for all checks to complete
+                bool[] results = await Task.WhenAll(checkTasks).ConfigureAwait(false);
+
+                // Remove failed endpoints from active list; record the failed endpoints
+                // for removal after releasing the active endpoints semaphore.
+                int endpointIndex = 0;
+                for (int resultIndex = 0; resultIndex < results.Length; resultIndex++)
+                {
+                    IEndpointInfo endpoint = _activeEndpoints[endpointIndex];
+                    if (results[resultIndex])
+                    {
+                        validEndpoints?.Add(endpoint);
+                        endpointIndex++;
+                    }
+                    else
+                    {
+                        _activeEndpoints.RemoveAt(endpointIndex);
+
+                        endpointsToRemove.Add(endpoint);
+                    }
+                }
+            }
+            finally
             {
-                throw new InvalidOperationException(nameof(ServerEndpointInfoSource.Start) + " method must be called before invoking this operation.");
+                _activeEndpointsSemaphore.Release();
+            }
+
+            if (endpointsToRemove.Count > 0)
+            {
+                lock (_pendingRemovalEndpoints)
+                {
+                    foreach (IEndpointInfo endpoint in endpointsToRemove)
+                    {
+                        _pendingRemovalEndpoints.Enqueue(endpoint);
```

Comment:
Rather than having a lock, queue, and semaphore for _pendingRemovalEndpoints, would it be easier to just make this a Channel that you dequeue from when data is available?